### PR TITLE
Give `modeling_t5.py` a `_prune_heads`

### PR DESCRIPTION
**Pull request by CaffreyR**
# What does this PR do?

 

Fixes # (issue) 19960 (not perfectly)
I refer a colab script in issue, it can prune but with forward problems. 
You can see here https://colab.research.google.com/drive/1b9mHjtn2UxuHU_Sb_RXts12rDzbebBX0#scrollTo=hUSe4a1oOp6D

I use `opendelta` to visualize the  pruning process.

But we seems to be a forward problem

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@ArthurZucker

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Great work! Fell free to ping me for a review 👍

**Comment by CaffreyR** [comment]
Hi @ArthurZucker , please pay more attention to the `position_bias`, I think I change it too sharply otherwise the shape will not be the same to the `score`

**Comment by CaffreyR** [comment]
Hi @ArthurZucker , I upload a new commits, it seems better deal with `position_bias`. And if we do not add `head_mask` and `decoder_head_mask` in the model `forward`, the code can run. But we just ignore this line's problem.

https://github.com/huggingface/transformers/blob/c3a93d8d821bc1df3601ba858e7385eada8db3a5/src/transformers/models/t5/modeling_t5.py#L547

**Comment by ArthurZucker** [comment]
Okay! That's way better, the least code we change the better. I remember seeing a similar fix, let me link that and I think we will be able to merge if slow tests pass 👍

**Comment by CaffreyR** [comment]
gently pin @patrickvonplaten Thanks

**Comment by ArthurZucker** [comment]
Hey, the tests are simply the CircleCI tests! Try running `make fixup` and `make fix-copies`. The integrations tests have to pass to be able to merge the PR!

**Comment by CaffreyR** [comment]
Hi @ArthurZucker  I have try to pull new request after `make fixup` and `make fix-copies`, but it still can not pass CircleCI :(

**Comment by ArthurZucker** [comment]
Hey! You didn't have to open a new PR! I will have a look and help you fix the tests ☺️

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -513,7 +513,7 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
         if position_bias is None:
             if not self.has_relative_attention_bias:
                 position_bias = torch.zeros(
-                    (1, self.n_heads, real_seq_length, key_length), device=scores.device, dtype=scores.dtype
+                    (1, self.n_heads+len(self.pruned_heads), real_seq_length, key_length), device=scores.device, dtype=scores.dtype
```
[the remark this answers is not in the document]
Comment (CaffreyR) [reply src/transformers/models/t5/modeling_t5.py:516]: Hi, in this line 

https://github.com/CaffreyR/transformers/blob/main/src/transformers/models/t5/modeling_t5.py#L369 

we have update the `self.n_heads`, but in the `forward`, we have to keep `position_bias` to remain `12`, so that we use mask to do the prune progress.

https://github.com/CaffreyR/transformers/blob/main/src/transformers/models/t5/modeling_t5.py#L531

**Comment by CaffreyR** [comment]
OK then ! Thanks! It seems my flat can not work on the `make fixup`

File: src/transformers/models/t5/modeling_t5.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (CaffreyR) [reply src/transformers/models/t5/modeling_t5.py:516]: You can see that after `prune_head` in T5 attention, the `self.n_heads` has already been the remaining heads number. But in the `forward` we have to tell the T5 attention the shape of `position_bias`

**Comment by ArthurZucker** [comment]
Okay, it seems that #19097 also wanted to adress part of this issue. Since it has not really progressed, we can do everything here

**Comment by CaffreyR** [comment]
Great! Anything that I can help?

**Comment by ArthurZucker** [comment]
OKay, try running : 
- `make style` to pass the `check_code_quality` 
- `git pull upstream main` to merge the changes from the `huggingface/transformers/main` branch. 
Then we will try to debug the tests that are failing. 
You can also do this by running `RUN_SLOW=1 pytest tests/models/t5/test_modeling_t5.py`.

**Comment by CaffreyR** [comment]
Hi @ArthurZucker ,I try to `make style`, it reports error 
```
All done! ✨ 🍰 ✨
597 files reformatted, 1299 files left unchanged.
isort examples tests src utils
Skipped 1 files
/Library/Developer/CommandLineTools/usr/bin/make autogenerate_code
running deps_table_update
updating src/transformers/dependency_versions_table.py
/Library/Developer/CommandLineTools/usr/bin/make extra_style_checks
python utils/custom_init_isort.py
python utils/sort_auto_mappings.py
doc-builder style src/transformers docs/source --max_len 119 --path_to_docs docs/source
make[1]: doc-builder: No such file or directory
make[1]: *** [extra_style_checks] Error 1
make: *** [style] Error 2

```

**Comment by CaffreyR** [comment]
Maybe it is my machine's problem of `make style`? @ArthurZucker  Macbook pro with m1 maybe?

**Comment by ArthurZucker** [comment]
No, the error is from the missing `huggingface_doc` package ! Don't worry. Try installing it. 
The files should still have been formatted

**Comment by CaffreyR** [comment]
```
(huggingface) caffrey@wangguorundeMacBook-Pro transformers % pip install huggingface_doc
ERROR: Could not find a version that satisfies the requirement huggingface_doc (from versions: none)
ERROR: No matching distribution found for huggingface_doc
```
I have tried it but it cannot be install on my machine! Is there any other way to do it?

**Comment by CaffreyR** [comment]
OK! I install it and rerun `make style` and `[python utils/check_copies.py --fix_and_overwrite](https://github.com/huggingface/transformers/pull/19975/commits/e6cb8a8c17eee85cb21ad4f1b5f8f717671257b0)`
Hope it will work this time

**Comment by CaffreyR** [comment]
Hi @ArthurZucker , this new version commits seems to be stuck in the test. It keeps waiting for several hours. What should I do?

**Comment by CaffreyR** [comment]
Hi @ArthurZucker it seems that `make style` can run well on linux but cannot run well in macOS system. :) Maybe it is better to find a new method for Apple M1 :)

**Comment by CaffreyR** [comment]
Hi @ArthurZucker , could you please give me a review? Many thanks!

**Comment by CaffreyR** [comment]
Hi @ArthurZucker

**Comment by ArthurZucker** [comment]
Hey, let's try to rebase to `be59316681fca13483da0ac2eac341f7df090e35`, since a loooot of files were modified by make style ( and this is not normal!). The issue most probably comes from your version of `black.  `pip install hf-doc-builder` or upgrading it sould solve this! 
I will review once that's clean! I can also help with make style if you are still unable to have the expected result! 🤗

**Comment by CaffreyR** [comment]
Hi @ArthurZucker , what is rebase? BTW?
